documentation/ZENKO-3397_CRUD_root_access_keys

wabernat opened the pull request:
Added API and command ref documentation for create-, list-, update-, and delete-access-key apis.

Also added "generate-access-key" documentation, though I am uncertain of its provenance. May be related to either "create-access-key" or "geerate-access-key-last-used" apis. Feel free to propose radical changes.

wabernat committed:
documentation/ZENKO-3397_CRUD_root_access_keys

Added API and command ref documentation for create-, list-, update-, and
delete-access-key apis.

Also added "generate-access-key" documentation, though I am uncertain of its
provenance. May be related to either "create-access-key" or
"geerate-access-key-last-used" apis. Feel free to propose radical changes.

wabernat committed:
documentation/ZENKO-3397_CRUD_root_access_keys

Added API and command ref documentation for create-, list-, update-, and
delete-access-key apis.

Also added "generate-access-key" documentation, though I am uncertain of its
provenance. May be related to either "create-access-key" or
"geerate-access-key-last-used" apis. Feel free to propose radical changes.

bert-e commented:
# Hello wabernatscality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/zenko/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-3397 contains:


* `2.0.0-alpha.8`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.0.0`


Please check the `Fix Version/s` of ZENKO-3397, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

lucieleonard commented:
/wait

wabernatScality commented:
/approve

anurag4DSB reviewed (commented):
I would suggest taking the latest one but only admin routes:
https://github.com/scality/Vault/blob/development/8.2/PROTOCOL.md

Meanwhile, I am also coming up with a command line

wabernat commented:
When you say "admin routes" are you referring to the routes described as "super admin routes" in the .md file? (list account, delete account, "all account routes")?

anurag4DSB commented:
> When you say "admin routes" are you referring to the routes described as "super admin routes" in the .md file? (list account, delete account, "all account routes")?

Yes, that is correct.

anurag4DSB reviewed (approved):
I understand the PR now. It is for CRUD for access keys.
Please feel free to ignore my previous comment.

wabernat commented:
Do you approve the technical content of this PR?

anurag4DSB commented:
@wabernatScality 
Here are the details: https://scality.atlassian.net/browse/ZENKO-3397?focusedCommentId=279016

cathydossantospinto committed:
ZENKO-3397 added topic on first accesskey creation using vaultclient

cathydossantospinto committed:
ZENKO-3397 added topic on first accesskey creation using vaultclient

anurag4DSB commented on docs/docsource/reference/apis/iam/create_access_key_using_vaultclient.rst:
```diff
@@ -0,0 +1,35 @@
+
+   .. code::
+
+      kubectl --kubeconfig=/etc/kubernetes/admin.conf -n zenko exec $VAULT_CONTAINER -- bash -c   "ADMIN_ACCESS_KEY_ID=${ADMIN_ACCESS_KEY_ID} ADMIN_SECRET_ACCESS_KEY=${ADMIN_SECRET_ACCESS_KEY} /vault/node_modules/vaultclient/bin/vaultclient create-account --name account2 --email account2@test.com --host 127.0.0.1  --port 8600"
```
On second thought we do not even need the port number of host. Its default.

```suggestion
      kubectl --kubeconfig=/etc/kubernetes/admin.conf -n zenko exec $VAULT_CONTAINER -- bash -c   "ADMIN_ACCESS_KEY_ID=${ADMIN_ACCESS_KEY_ID} ADMIN_SECRET_ACCESS_KEY=${ADMIN_SECRET_ACCESS_KEY} /vault/node_modules/vaultclient/bin/vaultclient create-account --name account1 --email account1@scality.local"
```

cathydossantospinto commented on docs/docsource/reference/apis/iam/create_access_key_using_vaultclient.rst:
```diff
@@ -0,0 +1,35 @@
+      ADMIN_ACCESS_KEY_ID=$(kubectl --kubeconfig=/etc/kubernetes/admin.conf  get secret artesca-data-management-vault-admin-creds.v1 -n zenko -o jsonpath='{.data.accessKey}' | base64 -d)
+      ADMIN_SECRET_ACCESS_KEY=$(kubectl --kubeconfig=/etc/kubernetes/admin.conf  get secret artesca-data-management-vault-admin-creds.v1 -n zenko -o jsonpath='{.data.secretKey}' | base64 -d)
+
+#. Create an account using ``vaulclient``.
```
Fix typo ``vaultclient``.

wabernat reviewed (commented):
Don't put this in the Reference. It belongs in Operation.

anurag4DSB commented on docs/docsource/reference/apis/iam/create_access_key_using_vaultclient.rst:
```diff
@@ -0,0 +1,35 @@
+#. Set the admin access key ID and secret access key.
+
+   .. code::
+
```
Before this I forgot this command:
```
    VAULT_CONTAINER=$(kubectl --kubeconfig=/etc/kubernetes/admin.conf  -n zenko get pods  -l app.kubernetes.io/name=connector-vault | sed -n '2 p' | awk '{print $1})'
```

miniscruff replied:
I am not going to say for sure but anytime I use kubectl + sed/awk there is a better option with `-o jsonpath`. Just a thought.

lucieleonard replied:
Is this command listing the pods?

anurag4DSB replied:
You are right, how about this?
```
    VAULT_CONTAINER=$(kubectl --kubeconfig=/etc/kubernetes/admin.conf -n zenko get pods -l app.kubernetes.io/name=connector-vault -o jsonpath='{.items[0].metadata.name}')
```

wabernat commented:
This is turning into two tickets: 
* Document the APIs relevant to generating a root access key (reference)
* Document how to generate a root access key (operation)
I would like to merge the reference material, then merge the HowTo stuff in another ticket. 

Does this seem reasonable, or are we out of time?

wabernat committed:
Moved first access key gen section to installation/configure/

Entered requested changes

wabernat committed:
Moved first access key gen section to installation/configure/

Entered requested changes

lucieleonard commented on docs/docsource/reference/apis/iam/create_access_key.rst:
```diff
@@ -0,0 +1,64 @@
+
+.. code::
+
+  POST /
```
```suggestion
   POST /
```

wabernat committed:
Update docs/docsource/reference/apis/iam/create_access_key.rst

Co-authored-by: Lucie Leonard <43170258+lucieleonard@users.noreply.github.com>

wabernat committed:
Update docs/docsource/reference/apis/iam/create_access_key.rst

Co-authored-by: Lucie Leonard <43170258+lucieleonard@users.noreply.github.com>

wabernat committed:
Merge branch 'documentation/ZENKO-3397_CRUD_root_access_keys' of github.com:scality/Zenko into documentation/ZENKO-3397_CRUD_root_access_keys

wabernat committed:
documentation/ZENKO-3397_CRUD_root_access_keys

Added api, cli, and use case info for create, list, and delete access keys.

Added phantom "update access key" docs (.txt format) to reference/CLI and APIs. Feature pending.

Added note to key commands concerning self-referential UserName option/field.

Updated Supported APIs.

Fixed IAM section.

anurag4DSB commented on docs/docsource/installation/configure/create_an_initial_access_key.rst:
```diff
@@ -0,0 +1,38 @@
+
+   .. code::
+
+      kubectl --kubeconfig=/etc/kubernetes/admin.conf -n zenko exec $VAULT_CONTAINER -- bash -c   "ADMIN_ACCESS_KEY_ID=${ADMIN_ACCESS_KEY_ID} ADMIN_SECRET_ACCESS_KEY=${ADMIN_SECRET_ACCESS_KEY} /vault/node_modules/vaultclient/bin/vaultclient generate-account-access-key --name=account2 --host 127.0.0.1  --port 8600"
```
In the first example, we are generating keys for `account1`, in generating account access keys we have `account2`. Not sure how much that matters, but thought it should be uniform.

wabernat replied:
changing to "account1"

anurag4DSB replied:
Thanks, we also need to remove --host and --port like other places.
Sorry but I thought I said it at one place so will be considered other places as well.

```suggestion
      kubectl --kubeconfig=/etc/kubernetes/admin.conf -n zenko exec $VAULT_CONTAINER -- bash -c   "ADMIN_ACCESS_KEY_ID=${ADMIN_ACCESS_KEY_ID} ADMIN_SECRET_ACCESS_KEY=${ADMIN_SECRET_ACCESS_KEY} /vault/node_modules/vaultclient/bin/vaultclient generate-account-access-key --name=account1"
```

anurag4DSB commented on docs/docsource/reference/apis/iam/create_access_key.rst:
```diff
@@ -0,0 +1,63 @@
+.. _CreateAccessKey:
+
+CreateAccessKey
```
Somehow we have to mention, if a username is not specified with all these calls:
* `CreateAccessKey`
* `UpdateAccessKey`
* `DeleteAccessKey`
* `ListAccessKeys`

If an admin makes the above API calls with account credentials without the parameter `UserName` then the action is on itself.
Similar for the user. If we use user credentials with proper policies to call create access key without a `UserName` parameter, it will create AK/SK for itself.

anurag4DSB replied:
Actually, it's mentioned below in the description of the AWS cli part.

wabernat replied:
I am reading this as "no action required". Please correct me if I misinterpret.

wabernat replied:
An unspecified username means self-reference. An unspecified user implementing this API is performing it on itself.

anurag4DSB replied:
That sounds right as discussed over google meet.

anurag4DSB commented on docs/docsource/reference/introduction/supported_apis.rst:
```diff
@@ -160,6 +160,30 @@ S3 API Endpoints
+   +---------------------------------------+----------------+------------+
+   | Operation Name                        | Operation Type | Available? |
+   +=======================================+================+============+
+   | :ref:`create-access-key`              | Vault (POST)   | Yes        |
```
Why is this vault and rest of them IAM?
Just curious.

wabernat replied:
I may be confused on this point (and please correct me if I am) I listed "create-access-key" as Vault because I could not find this API in the IAM documentation, and assumed we were stepping out of line with IAM. 

To be clear: We have a module named Vault. It implements the IAM protocol. What would be best from the user's perspective to list these under? IAM, Vault, or some combination of the two?

anurag4DSB replied:
`IAM` for all calls which are listed here: https://docs.aws.amazon.com/IAM/latest/APIReference/API_Operations.html

And for all other calls, we use vaultclient ie, admin API.

anurag4DSB commented on docs/docsource/reference/introduction/supported_apis.rst:
```diff
@@ -160,6 +160,30 @@ S3 API Endpoints
+   +---------------------------------------+----------------+------------+
+   | :ref:`list-access-keys`               | IAM (POST)     | Yes        |
+   +---------------------------------------+----------------+------------+
+   | :ref:`update-access-key`              | IAM (POST)     | Yes        |
```
@terrydervaux UpdateAccessKey is not supported yet. 
needs to be prioritized: https://scality.atlassian.net/browse/ZENKO-2825

wabernat replied:
Please let me know if you want me to delete this for 2.0.

anurag4DSB replied:
@terrydervaux I would let you decide this.

anurag4DSB reviewed (commented):
Overall LGTM.
Just some minor comments.

anurag4DSB commented on docs/docsource/installation/configure/create_an_initial_access_key.rst:
```diff
@@ -1,32 +1,35 @@
-- Admin credentials 
-- Admin routes
+- Admin credentials are available 
+- Admin routes are set
```
What does this mean?
How will end-user ensure that  Admin Routes are available? And what are admin routes over here?

wabernat replied:
I was looking at the logic of the sentence and making assumptions. If admin creds and routes are prerequisites, then they should be in this bullet list (with verbs, to parallel the structure of the first bullet point). If the purpose of this set of instructions is to create admin creds and routes, then these items do not belong in the Prerequisites section.

wabernat replied:
delete these two (admin creds + routes).

anurag4DSB commented on docs/docsource/installation/configure/create_an_initial_access_key.rst:
```diff
@@ -1,32 +1,35 @@
 - The data service is deployed
-- Admin credentials 
-- Admin routes
+- Admin credentials are available 
```
Admin credentials are being made available by the commands on line 25 and 26 on this same file.

wabernat replied:
- On the cluster on which Artesca is deployed.
 First set of keys... needs to be on that node.

anurag4DSB reviewed (approved):
LGTM, except for the pre-requisite part.(commented)

lucieleonard commented:
ping berte please merge

wabernat commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`



The following branches will **NOT** be impacted:

* `development/0.9`
* `development/1.0`
* `development/1.1`
* `development/1.2`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`



The following branches have **NOT** changed:

* `development/0.9`
* `development/1.0`
* `development/1.1`
* `development/1.2`



Please check the status of the associated issue ZENKO-3397.

Goodbye wabernatscality.
